src/scheduling/cost_models/simulated_quincy_cost_model.cc

```diff
@@ -0,0 +1,358 @@
+		double delta_preferred_machine,
+		double delta_preferred_rack,
+		Cost_t core_transfer_cost,
+		Cost_t tor_transfer_cost,
+		uint32_t percent_block_tolerance,
+		uint64_t machines_per_rack)
+  : resource_map_(resource_map),
+    job_map_(job_map),
+    task_map_(task_map),
+    knowledge_base_(kb),
+		proportion_machine_preferred_(delta_preferred_machine),
+		proportion_rack_preferred_(delta_preferred_rack),
+		core_transfer_cost_(core_transfer_cost),
+		tor_transfer_cost_(tor_transfer_cost),
+		percent_block_tolerance_(percent_block_tolerance),
+		machines_per_rack_(machines_per_rack),
+		filesystem_(dfs),
+		runtime_distribution_(runtime_distribution),
+		block_distribution_(block_distribution)
+		{
+  // Shut up unused warnings for now
+  CHECK_NOTNULL(leaf_res_ids);
+  CHECK_NOTNULL(task_bindings);
+
+  // initialise to a single, empty rack
+  rack_to_machine_map_.assign(1, std::list<ResourceID_t>());
+}
+
+SimulatedQuincyCostModel::~SimulatedQuincyCostModel() {
+  for (auto elt : preferred_machine_map_) {
+    ResourceCostMap_t *map = elt.second;
+    delete map;
+  }
+}
+
+// The cost of leaving a task unscheduled should be higher than the cost of
+// scheduling it.
+Cost_t SimulatedQuincyCostModel::TaskToUnscheduledAggCost(TaskID_t task_id) {
+	// TODO(adam): time dependent cost;
+  return 10000;
+}
+
+// The cost from the unscheduled to the sink is 0. Setting it to a value greater
+// than zero affects all the unscheduled tasks. It is better to affect the cost
+// of not running a task through the cost from the task to the unscheduled
+// aggregator.
+Cost_t SimulatedQuincyCostModel::UnscheduledAggToSinkCost(JobID_t job_id) {
+  return 0ULL;
+}
+
+// The cost from the task to the cluster aggregator models how expensive is a
+// task to run on any node in the cluster. The cost of the topology's arcs are
+// the same for all the tasks.
+Cost_t SimulatedQuincyCostModel::TaskToClusterAggCost(TaskID_t task_id) {
+	return cluster_aggregator_cost_[task_id];
+}
+
+Cost_t SimulatedQuincyCostModel::TaskToResourceNodeCost(TaskID_t task_id,
+                                               ResourceID_t resource_id) {
+  ResourceCostMap_t *cost_map = preferred_machine_map_[task_id];
+  return (*cost_map)[resource_id];
+}
+
+// Cost from machines to cores, always 0.
+Cost_t SimulatedQuincyCostModel::ResourceNodeToResourceNodeCost(
+    ResourceID_t source,
+    ResourceID_t destination) {
+  return 0ULL;
+}
+
+// The cost from the resource leaf to the sink is 0.
+Cost_t SimulatedQuincyCostModel::LeafResourceNodeToSinkCost(ResourceID_t resource_id) {
+  return 0ULL;
+}
+
+Cost_t SimulatedQuincyCostModel::TaskContinuationCost(TaskID_t task_id) {
+	// TODO(adam): task preemption support
+  return 0ULL;
+}
+
+Cost_t SimulatedQuincyCostModel::TaskPreemptionCost(TaskID_t task_id) {
+	// TODO(adam): task preemption support
+  return 0ULL;
+}
+
+Cost_t SimulatedQuincyCostModel::TaskToEquivClassAggregator(TaskID_t task_id,
```

Comment:
We discussed the reasoning behind this (a quick hack to allow direct mapping of tasks to racks, rather than via TECs). I'm not sure if we arrived at a conclusion what a better way of doing this would be? Needs more discussion